81 s3 tools

ylyangtw opened the pull request:
#81

valentinedwv commented on earthcube_utilities/src/ec/bucketutil.py:
```diff
@@ -0,0 +1,309 @@
+    @click.option('--cfgfile', help='gleaner config file', type=click.Path(exists=True))
+    @click.option('--s3server', help='s3 server address')
+    @click.option('--s3bucket', help='s3 bucket')
+    @click.option('--graphendpoint', help='graph endpoint')
```
I don't think we need a graphendpoint. this is just the s3 datastore.

ylyangtw replied:
Thanks! I had same opinion but forgot to remove it.

valentinedwv commented on earthcube_utilities/src/ec/bucketutil.py:
```diff
@@ -0,0 +1,309 @@
+@click.option('--milled', help='include milled', default=False)
+@click.option('--summon', help='include summon', default=True)
+@common_params
+def sourceurl(cfgfile, s3server, s3bucket, graphendpoint, upload, output, debug, url, summon, milled):
```
getting en error with cli like: sourceurl --cfgfile ./resources/testing/gleaner --url http://ds.iris.edu/ds/products/emtf/

ylyangtw replied:
Thanks for finding out this bug. I have changed s3bucket to ctx.bucket in sourceurl and other functions that have the same bugs.

valentinedwv commented on earthcube_utilities/src/ec/bucketutil.py:
```diff
@@ -0,0 +1,309 @@
+@click.option('--milled', help='include milled', default=False)
+@click.option('--summon', help='include summon', default=True)
+@common_params
+def duplicates(cfgfile, s3server, s3bucket, graphendpoint, upload, output, debug, summon, milled, path):
```
failed with cli: bucketutil.py duplicates --cfgfile ./resources/testing/gleaner

valentinedwv commented on earthcube_utilities/src/ec/bucketutil.py:
```diff
@@ -0,0 +1,309 @@
+
+@cli.command()
+@common_params
+def stats(cfgfile, s3server, s3bucket, graphendpoint, upload, output, debug):
```
ucketutil.py stats --cfgfile ./resources/testing/gleaner --source geocodes_demo_datasets 
Usage: bucketutil.py stats [OPTIONS]
Try 'bucketutil.py stats --help' for help.

Error: No such option: --source Did you mean --s3bucket?

ylyangtw replied:
added source param for stats

valentinedwv commented on earthcube_utilities/src/ec/bucketutil.py:
```diff
@@ -0,0 +1,309 @@
+@click.option('--milled', help='include milled', default=False)
+@click.option('--summon', help='include summon', default=True)
+@common_params
+def cull(cfgfile, s3server, s3bucket, graphendpoint, upload, output, debug, summon, milled, path):
```
cli: bucketutil.py cull --cfgfile ./resources/testing/gleaner --path summoned/geocodes_demo_datasets

ylyangtw replied:
fix

valentinedwv reviewed (changes requested):
Don't think we need graphendpoint.
Getting a few errors.
Also, can some of the outputs be done as text and csv as outlined in the issue.
Json is good for code, but lousy for human consumption.

know it was added later, but upload to the s3 store, and ouput to a named file, rather than just the console for some things.

valentinedwv commented:
duplicates... can we get source names in the output:
Url,Total,Examples
Url,Total,Examples
Url,Total,Examples
Url,Total,Examples
Url,Total,Examples

valentinedwv commented:
You can move a small portion of the logic for  the duplicates, and cull into methods in s3 class which will make them more easily testable.

The sourceUrl is a nice simple call, but  that would also make a good method.

ylyangtw commented:
> You can move a small portion of the logic for the duplicates, and cull into methods in s3 class which will make them more easily testable.
> 
> The sourceUrl is a nice simple call, but that would also make a good method.

I moved list duplicates to s3, so `bucketutil duplicates` only do the summary report and `bucketutil cull` only do the deletion.

valentinedwv commented:
oss.geocodes-aws-dev.earthcube.org/citest

Not all the files have needed X-Meta (due to an old bug)... so lets just catch them, and report them... note in doco that if duple is a single one, then it has missing metadata..

```
/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py duplicates --cfgfile ./resources/testing/gleaner 
Traceback (most recent call last):
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 300, in <module>
    res = cli()
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 59, in wrapper
    return func(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 215, in duplicates
    df = s3Minio.listDuplicates(ctx.bucket, path, summon, milled)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/datastore/s3.py", line 210, in listDuplicates
    path_to_run = path.strip("/")
AttributeError: 'NoneType' object has no attribute 'strip'
```

ylyangtw commented:
Thanks for the suggestions @valentinedwv ! I have made some changes:
- Added catch/try to codes using X-Meta
- Added `required=True` to some parameters, so click will handle them for us
- Changed urls output to csv. I tried json but csv makes more sense for reading, so I skipped implementing json.
I also briefly checked other methods' output formats, all make sense to me~

valentinedwv commented:
flight testing.
Cull seemed to work... went from 27 to 18... 

`bucketutil.py duplicates --cfgfile ./resources/testing/gleaner --path summmoned/geocodes_demo_datasets `

```
Traceback (most recent call last):
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3802, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 165, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 5745, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 5753, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'Url Duplicates'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 297, in <module>
    res = cli()
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 59, in wrapper
    return func(*args, **kwargs)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/earthcube_utilities/src/ec/bucketutil.py", line 213, in duplicates
    df = df[df['Url Duplicates'] > 1]
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/pandas/core/frame.py", line 3807, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/Users/valentin/development/dev_earthcube/earthcube_utilities/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3804, in get_loc
    raise KeyError(key) from err
KeyError: 'Url Duplicates'

Process finished with exit code 1

```

valentinedwv commented:
DUplicate not fully working.

Run url on geocodes_demo_datasets, you can see that url's are duplicated. Fragment

`bucketutil.py urls --cfgfile ./resources/testing/gleaner --source geocodes_demo_datasets `

```
sha,url
15fe96330ad8cefe9d39118b4ea2beb1702e0bd9.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/argo-20220707.json
2f4f53043c7dca1e2cc32ca1e5abf93fe627fa36.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/argoSimple-v1Shapes.json
50da29b7815edc772f6a0920973abe2e1254da13.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/bcodmo1.json
61af1c32ec3cef177fe6169132c3c9c4de62f284.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_iris_syngine.json
706852496ea174a9ccd5ac752cade0a3f45836a1.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_iris_syngine.json
8c30552eca65b8807d7fa6bb5e7b8517b93bb207.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/bcodmo1-20220707.json
9de034cf10657d9e4f395f6cc2eae9d2816a26fd.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_amgeo_data-01-06-2013-17-30-00.json
a340bc5e7139e9386176099c5a24e68db3923236.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/argoSimple-v1Shapes.json
ab96a594090ae7027c9b0f7a1460e74db771847c.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_lipdverse_HypkanaHajkova2016.json
b757fdf54ec4dec3211f1d04aa0f7f0e98c54be0.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/bcodmo1.json
d786b109bf94da2c14a0ad7b8c4afb428c710aca.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/opentopo1.json
d80cf2f5d7c80dc86af18141175edfeb8163ca2a.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_amgeo_data-01-06-2013-17-30-00.json
d82062e44ed98aa19718b2461a2a15c7c6dd51a3.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/earthchem_1572.json
df32d774997f575945eef2b7fdc2e20612e34779.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/argo-20220707.json
e48e13a7688098d5a5a9600068129505a37f484e.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/earthchem_1572.json
e6243fdf074d088b1e1215547e77ac4431067f89.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/MB_lipdverse_HypkanaHajkova2016.json
e6683000b60231254f0e1dc7af55f11943541beb.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/opentopo1.json
f2448275b659577ea9ecfaeee5c22f21ff3580fa.jsonld,https://raw.githubusercontent.com/earthcube/GeoCODES-Metadata/main/metadata/Dataset/allgood/bcodmo1-20220707.json
```

run duplicate, and nothing happens

```
bucketutil.py duplicates --cfgfile ./resources/testing/gleaner --path summmoned/geocodes_demo_datasets 
INFO : No object has been found

Process finished with exit code 0
```

ylyangtw commented:
duplicate work for me, it finds duplicate items...
![image](https://github.com/earthcube/earthcube_utilities/assets/63822845/177277ad-ba18-42c3-9002-f540a35a4d49)
